jalib/jassert.cpp

```diff
@@ -46,6 +43,9 @@ int jassert_quiet = 0;
 static int theLogFileFd = -1;
 static int errConsoleFd = -1;
 
+static jalib::string& tmpDir() {static jalib::string s;return s;};
+static jalib::string& uniquePidStr() {static jalib::string s;return s;};
```

Comment:
@gc00: Also, with multiple global (non-primitive) objects, the order in which their constructors are called is hard to reason about. It is not defined by the language specification. However, in this case, we want precise control over when the initialization of `tmpDir` happens. This is important because the DMTCP libraries, and the executables that statically link with `libjalib.a` can define their own global C++ objects. (In fact, libdmtcp relies on this trick to inject a ckpt thread in every process).

As @karya0 mentioned earlier, wrapping global C++ objects in functions is the idiomatic way of implementing this. This is portable and avoids non-deterministic behavior. See: https://isocpp.org/wiki/faq/ctors#static-init-order, and https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use.